mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1001,11 +996,18 @@ private[spark] object RandomForest extends Logging {
     } else {
       val numSplits = metadata.numSplits(featureIndex)
 
-      // get count for each distinct value
-      val (valueCountMap, numSamples) = featureSamples.foldLeft((Map.empty[Double, Int], 0)) {
+      // get count for each distinct value except zero value
+      val (partValueCountMap, partNumSamples) = featureSamples.foldLeft((Map.empty[Double, Int], 0)) {
         case ((m, cnt), x) =>
           (m + ((x, m.getOrElse(x, 0) + 1)), cnt + 1)
       }
+
+      // Calculate the number of samples for finding splits
+      val numSamples: Int = (samplesFractionForFindSplits(metadata) * metadata.numExamples).toInt
+
+      // add zero value count and get complete statistics
+      val valueCountMap: Map[Double, Int] = partValueCountMap + (0.0 -> (numSamples - partNumSamples))
```

Comment:
This also probably doesn't matter but won't the new (0.0, ...) element always come first? you could append it below after sorting the rest rather than add it earlier to the Map.